src/api/app/models/event/request.rb

```diff
@@ -77,16 +77,48 @@ def source_maintainers
       action_maintainers('sourceproject', 'sourcepackage')
     end
 
+    def source_watchers
+      source_or_target_project_watchers(project_type: 'sourceproject')
+    end
+
     def target_watchers
-      find_watchers('targetproject')
+      source_or_target_project_watchers(project_type: 'targetproject')
     end
 
-    def source_watchers
-      find_watchers('sourceproject')
+    def source_package_watchers
+      source_or_target_package_watchers(project_type: 'sourceproject', package_type: 'sourcepackage')
+    end
+
+    def target_package_watchers
+      source_or_target_package_watchers(project_type: 'targetproject', package_type: 'targetpackage')
     end
 
     private
 
+    def source_or_target_project_watchers(project_type:)
+      # TODO: remove old `find_watchers` implementation after rolling out `new_watchlist` from beta program
+      find_watchers(project_type) +
+        payload['actions'].map { |action| action[project_type] }
+                          .map { |project_name| Project.find_by_name(project_name) }
+                          .compact.map(&:watched_items)
+                          .flatten.map(&:user)
+    end
+
+    def source_or_target_package_watchers(project_type:, package_type:)
+      payload['actions'].map { |action| [action[project_type], action[package_type]] }
+                        .map do |project_name, package_name|
+        next if project_name.blank? || package_name.blank?
+
+        Package.get_by_project_and_name(project_name,
+                                        package_name,
+                                        { follow_multibuild: true, follow_project_links: false, use_source: false })
+      end
+                        .compact.map(&:watched_items)
+                        .flatten.map(&:user)
+    rescue Package::Errors::UnknownObjectError, Project::Errors::UnknownObjectError
+      []
+    end
```

Comment:
Shouldn't the `rescue` clause be moved just after the call to `get_by_project_and_name`? I'm thinking in a case (not yet confirmed) where several actions could contain different project/package combinations. The code, as it is now, will return `[]` just after the first ocurrence of project/package not found.

Reply:
Fixed, good catch!